code/_onclick/hud/action.dm

```diff
@@ -0,0 +1,174 @@
+				var/obj/effect/proc_holder/spell = target
+				spell.Click()
+		if(AB_INNATE)
+			if(!active)
+				Activate()
+			else
+				Deactivate()
+		if(AB_GENERIC)
+			if(target && procname)
+				call(target,procname)()
+	return
+
+/datum/action/proc/Activate()
+	return
+
+/datum/action/proc/Deactivate()
+	return
+
+/datum/action/proc/Process()
+	return
+
+/datum/action/proc/CheckRemoval(mob/living/user) // 1 if action is no longer valid for this mob and should be removed
+	return 0
+
+/datum/action/proc/IsAvailable()
+	return Checks()
+
+/datum/action/proc/Checks()// returns 1 if all checks pass
+	if(!owner)
+		return 0
+	if(check_flags & AB_CHECK_RESTRAINED)
+		if(owner.restrained())
+			return 0
+	if(check_flags & AB_CHECK_STUNNED)
+		if(owner.stunned)
+			return 0
+	if(check_flags & AB_CHECK_LYING)
+		if(owner.lying)
+			return 0
+	if(check_flags & AB_CHECK_ALIVE)
+		if(owner.stat)
+			return 0
+	if(check_flags & AB_CHECK_INSIDE)
+		if(!(target in owner))
+			return 0
+	return 1
+
+/datum/action/proc/GetIcon(var/button)
+	if(action_type == AB_ITEM && target)
+		var/obj/item/I = target
+		return image(I.icon, button , I.icon_state)
+	if(button_icon && button_icon_state)
+		return image(button_icon,button,button_icon_state)
+
+/obj/screen/action_button
+	var/datum/action/owner
+
+/obj/screen/action_button/Click()
+	if(usr.next_move >= world.time) // Is this needed ?
+		return
+	owner.Trigger()
+	return 1
+
+//Hide/Show Action Buttons ... Buttons
+
+/obj/screen/action_button_hide
+	icon = 'icons/mob/actions.dmi'
+	icon_state = "hide"
+
+/obj/screen/action_button_hide/Click()
+	usr.hud_used.action_buttons_hidden = 1
+	usr.client.screen -= src
+	world << usr.hud_used.ButtonNumberToScreenCoords(1)
+	world << usr.hud_used.ButtonNumberToScreenCoords(2)
+	world << usr.hud_used.ButtonNumberToScreenCoords(3)
+
+/obj/screen/action_button_show
+	icon = 'icons/mob/actions.dmi'
+	icon_state = "show"
+
+/obj/screen/action_button_show/Click()
+	usr.hud_used.action_buttons_hidden = 0
+	usr.client.screen -= src
+
+//This is the proc used to update all the action buttons. Properly defined in /mob/living/
+/mob/proc/update_action_buttons()
+	return
+
+//Presets for item actions
+/datum/action/item_action
+	check_flags = AB_CHECK_RESTRAINED|AB_CHECK_STUNNED|AB_CHECK_LYING|AB_CHECK_ALIVE|AB_CHECK_INSIDE
+
+/datum/action/item_action/CheckRemoval(mob/living/user)
+	return !(target in user)
+
+/datum/action/item_action/hands_free
+	check_flags = AB_CHECK_ALIVE|AB_CHECK_INSIDE
+
+#define AB_WEST_OFFSET 4
+#define AB_NORTH_OFFSET 26
+#define AB_MAX_COLUMNS 10
+
+/datum/hud/proc/ButtonNumberToScreenCoords(var/number) // TODO : Make this zero-indexed for readabilty
+	var/row = round(number/(AB_MAX_COLUMNS+1))
+	var/col = ((number - 1)%(AB_MAX_COLUMNS)) + 1
+	var/coord_col = col==1?"":"+[col-1]"
+	var/coord_col_offset = 4+2*col
+	var/coord_row = "[-1 - row]"
+	var/coord_row_offset = 26
+	return "WEST[coord_col]:[coord_col_offset],NORTH[coord_row]:[coord_row_offset]"
```

Comment:
Did that then found out screen loc and transform are not actually exclusive. Oh well